File: leetcode-js/198.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
    if (nums.length < 3) return Math.max(...nums);

    for (let i = 2; i < nums.length; i++) {
        nums[i] = Math.max(nums[i] + nums[i - 2], nums[i - 1]);
        nums[i - 1] = Math.max(nums[i - 1], nums[i - 2]);
    }
    return nums[nums.length - 1];
};

File: leetcode-go/198.go
package main

import (
	"math"
)

func rob(nums []int) int {
	for index, num := range nums {
		cacheBeforePrevHouse := num
		if index-2 >= 0 {
			cacheBeforePrevHouse += nums[index-2]
		}
		cacheInPrevHouse := 0
		if index-1 >= 0 {
			cacheInPrevHouse = nums[index-1]
		}
		nums[index] = int(math.Max(float64(cacheInPrevHouse), float64(cacheBeforePrevHouse)))
	}
	return nums[len(nums)-1]
}
